File: Samples/Verbum/how-to-handle-the-hyperlink-click-event-t437501-20.1.2-/CS/HyperlinkClickExample/Forms/SelectProductForm.cs
using DevExpress.XtraRichEdit.API.Native;
using System;
using System.Collections.Generic;
using System.Drawing;
using System.Windows.Forms;

namespace HyperlinkClickExample.Forms
{
    public partial class SelectProductForm : DevExpress.XtraEditors.XtraForm
    {
        #region #Properties
        object editValue;
        DocumentRange range;
        public virtual object EditValue { get { return editValue; } }
        public DocumentRange Range { get { return range; }  set { range = value; } }
        #endregion #Properties
        public SelectProductForm()
        {
            InitializeComponent();
        }

        #region #PopulateListBox
        //Populate the ListBoxControl with items based on the custom list retrieved from the main form:
        private void PopulateListBox(List<string> list)
        {
            this.listBoxControl.Items.AddRange(list.ToArray());
            if (list.Count > 0)
                this.listBoxControl.SelectedIndex = 0;
        }
        public SelectProductForm(List<string> list) : this()
        {
            PopulateListBox(list);
        }
        #endregion #PopulateListBox

        #region #Events
        //Handle the MouseMove and MouseClick events to support item selection by using mouse:
        private void listBoxControl_MouseMove(object sender, MouseEventArgs e)
        {
            for (int i = 0; i < this.listBoxControl.ItemCount; i++)
            {
                Rectangle rect = this.listBoxControl.GetItemRectangle(i);
                if (rect.Contains(e.X, e.Y))
                {
                    this.listBoxControl.SelectedIndex = i;
                    break;
                }
            }
        }
        private void listBoxControl_MouseClick(object sender, MouseEventArgs e)
        {
            if (e.Button != MouseButtons.Left || this.listBoxControl.SelectedIndex < 0)
                return;
            Rectangle rect = this.listBoxControl.GetItemRectangle(this.listBoxControl.SelectedIndex);
            if (!rect.Contains(e.X, e.Y))
                return;
            OnCommit();
        }
        //Handle the form KeyDown event to support keyboard navigation and item selection using the Enter key:
        private void SelectProductForm_KeyDown(object sender, KeyEventArgs e)
        {
            if (e.KeyData == Keys.Escape)
                Close();
            else if (e.KeyData == Keys.Enter)
                OnCommit();
        }
        protected virtual void OnCommit()
        {
            this.editValue = (string)this.listBoxControl.SelectedItem;
            if (onCommit != null)
                onCommit.Invoke(this, EventArgs.Empty);
            Close();
        }
        #endregion #Events

        #region #Commit
        EventHandler onCommit;
        public event EventHandler Commit { add { onCommit += value; } remove { onCommit -= value; } }
        #endregion #Commit
    }
}

File: Samples/Verbum/how-to-handle-the-hyperlink-click-event-t437501-20.1.2-/VB/HyperlinkClickExample/Forms/SelectProductForm.vb
Imports DevExpress.XtraRichEdit.API.Native
Imports System
Imports System.Collections.Generic
Imports System.Drawing
Imports System.Windows.Forms

Namespace HyperlinkClickExample.Forms
    Partial Public Class SelectProductForm
        Inherits DevExpress.XtraEditors.XtraForm

        #Region "#Properties"

        Private editValue_Renamed As Object

        Private range_Renamed As DocumentRange
        Public Overridable ReadOnly Property EditValue() As Object
            Get
                Return editValue_Renamed
            End Get
        End Property
        Public Property Range() As DocumentRange
            Get
                Return range_Renamed
            End Get
            Set(ByVal value As DocumentRange)
                range_Renamed = value
            End Set
        End Property
        #End Region ' #Properties
        Public Sub New()
            InitializeComponent()
        End Sub

        #Region "#PopulateListBox"
        'Populate the ListBoxControl with items based on the custom list retrieved from the main form:
        Private Sub PopulateListBox(ByVal list As List(Of String))
            Me.listBoxControl.Items.AddRange(list.ToArray())
            If list.Count > 0 Then
                Me.listBoxControl.SelectedIndex = 0
            End If
        End Sub
        Public Sub New(ByVal list As List(Of String))
            Me.New()
            PopulateListBox(list)
        End Sub
        #End Region ' #PopulateListBox

        #Region "#Events"
        'Handle the MouseMove and MouseClick events to support item selection by using mouse:
        Private Sub listBoxControl_MouseMove(ByVal sender As Object, ByVal e As MouseEventArgs) Handles listBoxControl.MouseMove
            For i As Integer = 0 To Me.listBoxControl.ItemCount - 1
                Dim rect As Rectangle = Me.listBoxControl.GetItemRectangle(i)
                If rect.Contains(e.X, e.Y) Then
                    Me.listBoxControl.SelectedIndex = i
                    Exit For
                End If
            Next i
        End Sub
        Private Sub listBoxControl_MouseClick(ByVal sender As Object, ByVal e As MouseEventArgs) Handles listBoxControl.MouseClick
            If e.Button <> MouseButtons.Left OrElse Me.listBoxControl.SelectedIndex < 0 Then
                Return
            End If
            Dim rect As Rectangle = Me.listBoxControl.GetItemRectangle(Me.listBoxControl.SelectedIndex)
            If Not rect.Contains(e.X, e.Y) Then
                Return
            End If
            OnCommit()
        End Sub
        'Handle the form KeyDown event to support keyboard navigation and item selection using the Enter key:
        Private Sub SelectProductForm_KeyDown(ByVal sender As Object, ByVal e As KeyEventArgs) Handles MyBase.KeyDown
            If e.KeyData = Keys.Escape Then
                Close()
            ElseIf e.KeyData = Keys.Enter Then
                OnCommit()
            End If
        End Sub
        Protected Overridable Sub OnCommit()
            Me.editValue_Renamed = CStr(Me.listBoxControl.SelectedItem)
            If Not onCommit_RenamedEvent Is Nothing Then
                RaiseEvent onCommit_Renamed(Me, EventArgs.Empty)
            End If
            Close()
        End Sub
#End Region ' #Events

#Region "#Commit"

        Private Event onCommit_Renamed As EventHandler
        Public Custom Event Commit As EventHandler
            AddHandler(ByVal value As EventHandler)
                AddHandler onCommit_Renamed, value
            End AddHandler
            RemoveHandler(ByVal value As EventHandler)
                RemoveHandler onCommit_Renamed, value
            End RemoveHandler
            RaiseEvent(ByVal sender As System.Object, ByVal e As System.EventArgs)
            End RaiseEvent
        End Event
#End Region ' #Commit
    End Class
End Namespace





